### ISSUE TITLE
How to keep OOV words in the translated text instead of showing UNK?

### ISSUE THREAD

**Issue by anglil** [tensorflow/tensor2tensor#387]
When BPE is not used, there could be a lot of OOV words not covered by training data. In this case, how to keep OOV words in the translated text (e.g., dev, test), instead of showing UNK? i.e., leave the untranslated word as is?

**Comment by martinpopel** [comment]
The translation problems in T2T use either BPE or "word-piece" subwords. Both techniques are very similar. There is no need for handling OOV with subwords. If you don't want to use subwords (but why?), there are several ways described in the literature for handling OOVs, e.g. fallback SMT phrase table, aligning OOVs in a sentence and leaving untranslated, or hybrid word+character model (#384).

**Comment by anglil** [comment]
Thanks. I'm empirically studying with certain language pairs, whether BPE or whole-word-tokens is a better way to go. And when taking the whole-word-tokens approach, what happens is that in the translated text there are lots of OOVs, labeled as UNK (or something else one specifically designates), and instead of showing these OOVs with the label in the translated text, I was wondering if T2T can exactly pinpoint which OOV corresponds to which untranslated word, either by alignment from the attention table or falling back to phrase-based tables? Just whether there is a flag in T2T that can keep OOVs as is in the translated text (as opposed to showing labels (e.g. UNK) in the translated text)?

**Comment by lukaszkaiser** [comment]
It isn't clear how to pinpoint which UNK/OOV is which word. There were techniques for that (see https://arxiv.org/abs/1410.8206) but nowadays BPE/WP=Subwords is the dominant approach as it is both easier and yields better results (see Table 4 and 5 in https://arxiv.org/abs/1609.08144). There is nothing inherent in T2T that can discover which OOV fits where, you'd need to re-implement the methods above to see for yourself.
